### PR TITLE
Support StatefulSets with 0 replicas

### PR DESCRIPTION
**Pull request by dturn**
**What are you trying to accomplish with this PR?**

Currently when one deploys a StatefulSet + Service pair where the SS's replicas is set to zero, the service will eventually time out. This behavior is wrong, its valid for a service to point at a SS that's intentionally been scaled down. We should treat this case the same as a deployment. 

fixes: https://github.com/Shopify/kubernetes-deploy/issues/265

**How is this accomplished?**
Have the service check to see if it's actually pointing at a deployment or statefulset that has been scaled down. 

**What could go wrong?**

- We're intentionally not handling the case of a service in front of other pod generating controllers: CronJob/Job.  The assumption here is its unusual to point a service at  collection of pods that normally isn't running. Further since they normally aren't running, I'd have expected a bug report about it by now.

- This keeps the same bad assumption that the pod selector will also select the deployment / SS that was raised in https://github.com/Shopify/kubernetes-deploy/issues/483. I've got a second PR to solve it https://github.com/Shopify/kubernetes-deploy/pull/541 so feel ok doing this in two steps.

### CODE REVIEW

File: lib/kubernetes-deploy/kubernetes_resource/service.rb [hunk 1]
```diff
@@ -7,9 +7,11 @@ def sync(cache)
       super
       if exists? && selector.present?
         @related_deployments = cache.get_all(Deployment.kind, selector)
+        @related_statefulsets = cache.get_all(StatefulSet.kind, selector)
```
Comment (timothysmith0609) [inline lib/kubernetes-deploy/kubernetes_resource/service.rb]: I think this is fine for now, but if we find ourselves having to accommodate some other kind of workload we should consider aggregating things into a `@related_workloads = {}` to simplify processing

Comment (dturn) [reply lib/kubernetes-deploy/kubernetes_resource/service.rb]: Yep, tackling that in https://github.com/Shopify/kubernetes-deploy/pull/541/files#diff-bfab9900d04f0a61190aa04f896aed1fR10

Comment (timothysmith0609) [reply lib/kubernetes-deploy/kubernetes_resource/service.rb]: :+1: Yep just saw that and was on my way to delete this comment

**Comment by dturn** [comment]
>Should we warn when a service is targeting a workload with (explicitly-set) 0 replicas, or do we trust the user to know what's going on?

We've got:
```
 Service/stateful-busybox Doesn't require any endpoint
 StatefulSet/stateful-busybox 0 replicas
```
in the output already, that seems like enough to me.

>The timeout message currently states: This service does not seem to select any pods. This means its spec.selector is probably incorrect.. This is not necessarily the timeout condition, though. What we want to say is that no pods are selected but there should be.


How about:

"This service does not seem to select any pods. This is likely invalid please confirm the spec.selector is correct and the targeted workload is healthy."

File: lib/kubernetes-deploy/kubernetes_resource/service.rb [hunk 2]
```diff
@@ -30,20 +32,22 @@ def deploy_succeeded?
...
     end
 
     def deploy_failed?
       false
     end
 
     def timeout_message
-      "This service does not seem to select any pods. This means its spec.selector is probably incorrect."
+      "This service does not seem to select any pods. "\
+      "This is likely invalid please confirm the spec.selector is correct and "\
```
Comment (RyanBrushett) [inline lib/kubernetes-deploy/kubernetes_resource/service.rb]:
```suggestion
      "This is likely invalid. Please confirm the spec.selector is correct and "\
```

⛏

**Pull request merged**